src/lib.rs

```diff
@@ -187,20 +199,39 @@ impl<D> EventSource for WaylandSource<D> {
             }
         }
 
-        // ensure we are not waiting for messages that are already in the buffer.
-        Self::loop_callback_pending(&mut self.queue, &mut callback)?;
-        self.read_guard = Some(Self::prepare_read(&mut self.queue)?);
-
-        Ok(())
+        // TODO: ensure we are not waiting for messages that are already in the buffer.
+        // is this needed? I think in the case that this would occur, getting the guard
+        // would fail anyway
+        // Self::loop_callback_pending(&mut self.queue, &mut callback)?;
+        self.read_guard = self.queue.prepare_read();
```

Comment:
Indeed, `prepare_read()` will return `None` if there are messages in the buffer, that is the exact condition for getting the guard to fail.